File: LeetcodeEasyProblems/_190. Reverse Bits/Solution.py
class Solution:
    # @param n, an integer
    # @return an integer
    def reverseBits(self, n):
        # Define the result bit
        res = 0
        # Loop 32 times over each bit in n
        for i in range (32):
            # Shift the result bit to the left by 1
            # creating space for the next bit
            res <<= 1
            # Using the bit function AND to take the 
            #least significant bit (LSB) of n
            new_bit = (n & 1)
            # Using the bit function OR to take in the new bit 
            # and put it at the right side of res
            res |= new_bit
            # Shift the n bit to the right by 1
            # change its LSB to the next bit
            n >>= 1
        
        return res

        

File: LeetcodeEasyProblems/_190. Reverse Bits/Solution.java
public class Solution {
    // you need treat n as an unsigned value
    public static int reverseBits(int n) {
        //Turn the integer n into the String for ease of changing it
        String bitInteger = String.valueOf(n);
        //If the entered number is not an 32 bit integer, return -1
        if (bitInteger.length() != 32){
            return -1;
        }
        //Create a placeholder character and an empty string
        char placeholderChar;
        String  reversedBitInteger = "";
        // The for loop to take each character in the original string 
        // mirror it through the central axis 
        for (int i = 31; i>=0;i--){
            placeholderChar = bitInteger.charAt(i);
            reversedBitInteger += placeholderChar;
        }
        // Convert the bit integer to its decimal counterpart and return it
        int reversedDecimalInteger = 0;
        for (int i = 31; i >=0; i --){
            placeholderChar = reversedBitInteger.charAt(i);
            if (i > 0) {
                reversedDecimalInteger+= (int)Math.pow(2,(31-i)) * Character.getNumericValue(placeholderChar);
            } else {
                // When the most significant digit of the bit is 1, then the sign is negative
                if (placeholderChar == 1){
                    reversedDecimalInteger -= (int)Math.pow(2,31);
                }
            }

        }
        return reversedDecimalInteger;
    }

    public static void main(String[] args) {
        System.out.println(reverseBits(00000010100101000001111010011100));
    }
}